src/pal/src/cruntime/misctls.cpp

```diff
@@ -150,250 +164,155 @@ See MSDN for more information.
+        return lpStartOfReturnBuffer;
+    } 
+    
+    int tempBufferLength = snprintf(TempBuffer, ECVT_MAX_BUFFER_SIZE, "%.40e", value);
+    _ASSERTE(tempBufferLength > 0 && ECVT_MAX_BUFFER_SIZE > tempBufferLength);
+    
+    //
+    // Calculate the exponent value
+    //
 
-        *dec = 1;
-        if ( *TempBuffer == '-' )
-        {
-            *sign = 1;
-        }
-        goto done;
+    int exponentIndex = tempBufferLength - 1;
+    while (TempBuffer[exponentIndex] != 'e' && exponentIndex > 0)
+    {
+        exponentIndex--;
     }
 
-    nTempBufferLength = strlen( TempBuffer );
-    lpEndOfTempBuffer = &(TempBuffer[ nTempBufferLength ]);
+    _ASSERTE(exponentIndex > 0 && (exponentIndex < tempBufferLength - 1));
 
-    /* Extract the exponent, and convert it to integer. */
-    while ( *lpEndOfTempBuffer != 'e' && nTempBufferLength > 0 )
+    int i = exponentIndex + 1;
+    int exponentSign = 1;
+    if (TempBuffer[i] == '-')
     {
-        nTempBufferLength--;
-        lpEndOfTempBuffer--;
+        exponentSign = -1;
+        i++;
     }
-    
-    ExponentBuffer[ 0 ] = '\0';
-    if (strncat_s( ExponentBuffer, sizeof(ExponentBuffer), lpEndOfTempBuffer + 1, 5 ) != SAFECRT_SUCCESS)
+    else if (TempBuffer[i] == '+')
     {
-        ERROR( "strncat_s failed!\n" );
-        *lpStartOfReturnBuffer = '\0';
-        goto done;
+        i++;
     }
 
-    nExponentValue = atoi( ExponentBuffer );
-
-    /* End the string at the 'e' */
-    *lpEndOfTempBuffer = '\0';
-    nTempBufferLength--;
+    int exponentValue = 0;
+    while (i < tempBufferLength)
+    {
+        _ASSERTE(TempBuffer[i] >= '0' && TempBuffer[i] <= '9');
+        exponentValue = exponentValue * 10 + ((BYTE) TempBuffer[i] - (BYTE) '0');
+        i++;
+    }
+    exponentValue *= exponentSign;
+    
+    //
+    // Determine decimal location.
+    // 
 
-    /* Determine decimal location. */
-    if ( nExponentValue == 0 )
+    if (exponentValue == 0)
     {
         *dec = 1;
     }
     else
     {
-        *dec = nExponentValue + 1;
-    }
-
-    if ( value == 0.0 )
-    {
-        *dec = 0;
+        *dec = exponentValue + 1;
     }
-    /* Copy the string from the temp buffer upto count characters, 
-    removing the sign, and decimal as required. */
-    lpTempBuffer = TempBuffer;
-    *lpReturnBuffer = '0';
-    lpReturnBuffer++;
+    
+    //
+    // Copy the string from the temp buffer upto precision characters, removing the sign, and decimal as required.
+    // 
 
-    while ( LoopIndex < ECVT_MAX_COUNT_SIZE )
+    i = 0;
+    int mantissaIndex = 0;
+    while (i < count && mantissaIndex < exponentIndex)
     {
-        if ( isdigit(*lpTempBuffer) )
-        {
-            *lpReturnBuffer = *lpTempBuffer;
-            LoopIndex++;
-            lpReturnBuffer++;
-        }
-        lpTempBuffer++;
-
-        if ( LoopIndex == count + 1 )
+        if (TempBuffer[mantissaIndex] >= '0' && TempBuffer[mantissaIndex] <= '9')
```

Comment:
we need to avoid copying the sign and the decimal point. so I have to check the numbers